### PR TITLE
TMDM-11733 beforeSaving process doesn't work if removing field of record

### PR DESCRIPTION
**Pull request by zfkang-talend**
JIRA: https://jira.talendforge.org/browse/TMDM-11733

**What is the current behavior?** (You can also link to an open issue here)
If one before saving process contains one job which it contains item, can't update the data from WEBUI. because when execute ApplyAction, the databaseDocument had been updated and can not commit to database, then will invoke class BeforeSaving, in this class, will invoke the method invokeBeforeSaving to execute the before saving process job, the parameter is databaseDocument.exportToString(), here databseDocument had been updated and return value item is the modified, here, the return item is not null, used the return item(modified value) as a userDocument to execute save follow, the modified value compare the value from database, have no change, at last can not update the value.

**What is the new behavior?**
If one before saving process contains one job which it contains item, when executed to ApplyAction, can't clean the filed value is null from fieldToValue map, add a new class to generate the string value of the record(include field vaue is null), use this string value as a parameter of invokeBeforeSaving method. for this modify, have two import modified:

1.  In DataRecord class, add one function:isExixedInValueList to judge the field is existed in DataRecord fieldToValue list.
2.  Add one class DataRecordEmptyLabelXmlWriter to generate the string value, if the field is existed in fieldToValue list, but the value is null, will generate an empty label, like this:<name/>

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecord.java [hunk 1]
```diff
@@ -203,7 +203,25 @@ public Object get(String fieldName) {
         }
         return currentValue;
     }
-    
+
+    public boolean isExixedInValueList(FieldMetadata field) {
```
Comment (pbabin) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecord.java]: Don't know why it specifies ValueList as it does not mandate it has to be a list.
a better naming would be `containsFieldToValue`

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecord.java [hunk 2]
```diff
@@ -203,7 +203,25 @@ public Object get(String fieldName) {
...
+        if (field == null) {
+            return false;
+        }
+        return fieldToValue.containsKey(field);
+    }
+
+    public boolean isExixedInValueList(String fieldName) {
```
Comment (pbabin) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecord.java]: `containsFieldToValue`

**Comment by pbabin** [comment]
Update PR descriptions according to the changes, and as well it would be godd to add the JIRA link so that immediately we can jump onto the corresponding JIRA

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java [hunk 3]
```diff
@@ -154,6 +259,62 @@ public void testMultiContainedType() throws Exception {
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithMultiContained><Id>ABCD</Id><Contained><ContainedId>CID1</ContainedId></Contained><Contained><ContainedId>CID2</ContainedId><ContainedName>CName2</ContainedName></Contained></WithMultiContained>",
+                result);
+    }
+
+    @Test
+    public void testMultiContainedTypeContainsNull2() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: This case is not necessary.

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java [hunk 4]
```diff
@@ -154,6 +259,62 @@ public void testMultiContainedType() throws Exception {
                 result);
     }
 
+    @Test
+    public void testMultiContainedTypeContainsNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: This case is not necessary.

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java [hunk 5]
```diff
@@ -125,6 +186,50 @@ public void testContainedType() throws Exception {
...
+        ContainedTypeFieldMetadata containedField = (ContainedTypeFieldMetadata) field;
+        ContainedComplexTypeMetadata tm = (ContainedComplexTypeMetadata) containedField.getType();
+
+        DataRecord contained = createDataRecord(tm);
+
+        setDataRecordField(record, "Contained", contained);
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals("<WithContained><Id>ABCD</Id><Contained></Contained></WithContained>", result);
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: `<Contained></Contained>` should not exist

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java [hunk 6]
```diff
@@ -101,6 +123,45 @@ public void testTypeWithComplexTypeOfReference() throws Exception {
...
+        setDataRecordField(record, "Name", null);
+
+        FieldMetadata field = type.getField("Address");
+
+        DataRecord referenced = createDataRecord(((ContainedComplexTypeMetadata) field.getType()).getContainedType());
+
+        setDataRecordField(record, "Address", referenced);
+
+        String result = toXmlString(record);
+        Assert.assertEquals("<Customer><Id>12345</Id><Address></Address></Customer>", result);
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: Should `<Address></Address>` exist?

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java (hunk 5 above)
Comment (zfkang-talend) [reply org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: If the file type is ReferenceFieldMetadata or ContainedTypeFieldMetadata, when the field value is null or empty, always generate a empty node, like '<Contained></Contained>'.

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java (hunk 6 above)
Comment (zfkang-talend) [reply org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: If the file type is ReferenceFieldMetadata or ContainedTypeFieldMetadata, when the field value is null or empty, always generate a empty node, like '<Contained></Contained>'.

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    public DataRecordContainsNullValueXmlWriter(String rootElementName) {
+        super(rootElementName);
+    }
+
+    public DataRecordContainsNullValueXmlWriter(ComplexTypeMetadata type) {
+        super(type);
+    }
+
+    @Override
+    public void write(DataRecord record, Writer writer) throws IOException {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java]: Try to refactor to reuse the common code with `DataRecordXmlWriter `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithContained><Id>ABCD</Id><Contained><ContainedId>CID</ContainedId><ContainedName>CName</ContainedName></Contained></WithContained>",
+                result);
+    }
+
+    @Test
+    public void testContainedTypeContainsNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testContainedTypeWithNullField `

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        DefaultMetadataVisitor<Void> fieldPrinter = new ContainsNullValueFieldPrinter(record, writer);
+        write(record, writer, fieldPrinter);
+    }
+
+    class ContainsNullValueFieldPrinter extends DataRecordXmlWriter.FieldPrinter {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java]: Rename to `FieldWithNullValuePrinter`

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+import org.talend.mdm.commmon.metadata.DefaultMetadataVisitor;
+import org.talend.mdm.commmon.metadata.EnumerationFieldMetadata;
+import org.talend.mdm.commmon.metadata.FieldMetadata;
+import org.talend.mdm.commmon.metadata.ReferenceFieldMetadata;
+import org.talend.mdm.commmon.metadata.SimpleTypeFieldMetadata;
+
+import com.amalto.core.storage.StorageMetadataUtils;
+import com.amalto.core.storage.record.metadata.DataRecordMetadata;
+
+public class DataRecordContainsNullValueXmlWriter extends DataRecordXmlWriter {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriter.java]: Rename to `DataRecordWithNullFieldsXmlWriter `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    @Before
+    public void setup() throws Exception {
+        super.setup();
+        writer = new DataRecordContainsNullValueXmlWriter();
+        writer.setSecurityDelegator(delegate);
+        repository.load(this.getClass().getResourceAsStream("metadata.xsd"));
+    }
+
+    @Test
+    public void testSimpleComplexType() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexType`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+        setDataRecordField(record, "Description", "Desc");
+        setDataRecordField(record, "Availability", Boolean.FALSE);
+        String result = toXmlString(record);
+        Assert.assertEquals(
+                "<SimpleProduct><Id>12345</Id><Name>Name</Name><Description>Desc</Description><Availability>false</Availability></SimpleProduct>",
+                result);
+    }
+
+    @Test
+    public void testSimpleTypeWithArray() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithArray`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+        DataRecord record = createDataRecord(repository.getComplexType("WithArray"));
+        setDataRecordField(record, "Id", "12345");
+        setDataRecordField(record, "Repeat", "ABC");
+        setDataRecordField(record, "Repeat", "DEF");
+        String result = toXmlString(record);
+        Assert.assertEquals("<WithArray><Id>12345</Id><Repeat>ABC</Repeat><Repeat>DEF</Repeat></WithArray>", result);
+    }
+
+    @Test
+    public void testSimpleTypeWithArrayContainsNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithArrayContainsNull `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+        DataRecord record = createDataRecord(repository.getComplexType("WithArray"));
+        setDataRecordField(record, "Id", "12345");
+        setDataRecordField(record, "Repeat", "ABC");
+        setDataRecordField(record, "Repeat", null);
+        String result = toXmlString(record);
+        Assert.assertEquals("<WithArray><Id>12345</Id><Repeat>ABC</Repeat><Repeat/></WithArray>", result);
+    }
+
+    @Test
+    public void testSimpleTypeWithEmptyArray() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithEmptyArray `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        DataRecord record = createDataRecord(repository.getComplexType("WithArray"));
+        setDataRecordField(record, "Id", "12345");
+        setDataRecordField(record, "Repeat", new ArrayList<String>());
+        String result = toXmlString(record);
+        Assert.assertEquals("<WithArray><Id>12345</Id><Repeat/></WithArray>", result);
+    }
+
+    @Test
+    public void testSimpleComplexTypeWithNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeContainsNull `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+        setDataRecordField(record, "Name", null);
+        setDataRecordField(record, "Description", "Desc");
+        setDataRecordField(record, "Availability", null);
+        String result = toXmlString(record);
+        Assert.assertEquals("<SimpleProduct><Id>12345</Id><Name/><Description>Desc</Description><Availability/></SimpleProduct>",
+                result);
+    }
+
+    @Test
+    public void testTypeWithComplexTypeOfReference() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithReferenceField `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        setDataRecordField(record, "Address", referenced);
+
+        String result = toXmlString(record);
+        Assert.assertEquals(
+                "<Customer><Id>12345</Id><Name>Name</Name><Address><Id>AddressId</Id><Street>AddressStreet</Street><City>AddressCity</City></Address></Customer>",
+                result);
+    }
+
+    @Test
+    public void testTypeWithComplexTypeOfReferenceContainsNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithReferenceContainsNull `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+        setDataRecordField(record, "Address", referenced);
+
+        String result = toXmlString(record);
+        Assert.assertEquals(
+                "<Customer><Id>12345</Id><Name/><Address><Id>AddressId</Id><Street>AddressStreet</Street><City/></Address></Customer>",
+                result);
+    }
+
+    @Test
+    public void testTypeWithComplexTypeOfReferenceContainsNullReference() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithEmptyReference`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+        DataRecord referenced = createDataRecord(((ContainedComplexTypeMetadata) field.getType()).getContainedType());
+
+        setDataRecordField(record, "Address", referenced);
+
+        String result = toXmlString(record);
+        Assert.assertEquals("<Customer><Id>12345</Id><Name/><Address></Address></Customer>", result);
+    }
+
+    @Test
+    public void testContainedType() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithContainedField`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithContained><Id>ABCD</Id><Contained><ContainedId>CID</ContainedId><ContainedName/></Contained></WithContained>",
+                result);
+    }
+
+    @Test
+    public void testNullContainedType() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithNullContainedField`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+
+        setDataRecordField(record, "Contained", contained);
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals("<WithContained><Id>ABCD</Id><Contained></Contained></WithContained>", result);
+    }
+
+    @Test
+    public void testMultiContainedType() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithMultiContainedField`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithMultiContained><Id>ABCD</Id><Contained><ContainedId>CID1</ContainedId><ContainedName>CName1</ContainedName></Contained><Contained><ContainedId>CID2</ContainedId><ContainedName>CName2</ContainedName></Contained></WithMultiContained>",
+                result);
+    }
+
+    @Test
+    public void testMultiContainedTypeContainsNull() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithMultiContainedFieldContainsNull `

Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: This case can be removed

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithMultiContained><Id>ABCD</Id><Contained><ContainedId>CID1</ContainedId><ContainedName/></Contained><Contained><ContainedId>CID2</ContainedId><ContainedName>CName2</ContainedName></Contained></WithMultiContained>",
+                result);
+    }
+
+    @Test
+    public void testMultiContainedTypeContainsNull2() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithMultiContainedFieldContainsNull`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals(
+                "<WithMultiContained><Id>ABCD</Id><Contained><ContainedId>CID1</ContainedId><ContainedName/></Contained><Contained></Contained></WithMultiContained>",
+                result);
+    }
+
+    @Test
+    public void testMultiContainedTypeWithEmptyList() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithEmptyMultiContainedField `

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+
+        setDataRecordField(record, "Contained", new ArrayList<DataRecord>());
+
+        String result = toXmlString(record);
+
+        Assert.assertEquals("<WithMultiContained><Id>ABCD</Id><Contained/></WithMultiContained>", result);
+    }
+
+    @Test
+    public void testEnumType() throws Exception {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordContainsNullValueXmlWriterTestCase.java]: Rename to `testComplexTypeWithEnum`

File: org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java [hunk 26]
```diff
@@ -26,7 +26,7 @@
 
 import com.amalto.core.load.io.ResettableStringWriter;
 
-public class DataRecordXmlWriterTestCase extends DataRecordDataWriterTestCase{
+public class DataRecordXmlWriterTestCase extends DataRecordDataWriterTestCase {
```
Comment (lpw82gf) [inline org.talend.mdm.core/test/com/amalto/core/storage/record/DataRecordXmlWriterTestCase.java]: Rename all the case names as used in`DataRecordContainsNullValueXmlWriterTestCase `

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordWithNullFieldsXmlWriter.java [hunk 27]
```diff
@@ -0,0 +1,240 @@
...
+                        out.write("</" + enumField.getName() + ">"); //$NON-NLS-1$ //$NON-NLS-2$
+                    } else {
+                        List valueAsList = (List) value;
+                        for (Object currentValue : valueAsList) {
+                            if (currentValue != null) {
+                                out.write("<" + enumField.getName() + ">"); //$NON-NLS-1$ //$NON-NLS-2$
+                                handleSimpleValue(enumField, currentValue);
+                                out.write("</" + enumField.getName() + ">"); //$NON-NLS-1$ //$NON-NLS-2$
+                            }
+                            if (currentValue != null) {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordWithNullFieldsXmlWriter.java]: Should ` if (currentValue != null) {` be `else {`

**Pull request merged**